### ISSUE TITLE
Federated horizontal result does not align with basic training without federation

### ISSUE THREAD

**Issue by ZiyueXu77** [dmlc/xgboost#10248]
This may not be a "bug", could be an "expected behavior" under our current settings.

I made a dataset with a fixed training / validation split, then further split the training into 3 parts, one for each horizontal client (see data preparation in [this example](https://github.com/ZiyueXu77/NVFlare/tree/secureboost/examples/advanced/xgboost_secure))

Then I run two options:
- basic training with the training (without split) and validation 
- horizontal federated training with the 3 splits and the same validation set

The observation is that the outcome histograms and resulting models are different, while in theory given perfect global sync information, the two should be equivalent.

My hypothesis is that the local histograms, since computed based on local data, would be different across clients, since partial histograms may not contain sufficient information to precisely recover the global histogrma, this can lead to the discrepancy during syncing such information and producing the global version. 

More investigation may be needed to find the root cause to tell if this is something expected to happen and thus could be a limitation of horizontal federation, which was assumed to be "lossless".

**Comment by ZiyueXu77** [comment]
Additional information: did some print outs over several places of "quantile.cc", and the cuts results are indeed different slightly:
Baseline: -56.4075 -20.3518 -16.3087 -13.7437 -12.4456 -11.3925 -10.7218 -10.0299 -9.65642 -9.21496
v.s. 
Horizontal: -56.4075 -20.785 -16.4087 -13.8562 -12.4785 -11.2658 -10.6562 -10.2007 -9.7461 -9.22402 

I think it is because the local histograms does not hold sufficient info to fully recover

**Comment by rongou** [comment]
Could these just be rounding errors with `float`s? I don't know how easy it is, perhaps you can try to replace `GradientPair` with `GradientPairPrecise` in the code and see it that helps.

**Comment by ZiyueXu77** [comment]
These are not gradients, they are the quantile cut values. 
As quantile stats does not have a simple algorithm for distributed aggregation, I guess here we did not use exact quantile algorithm, instead, approximate quantile is being used.

**Comment by ZiyueXu77** [comment]
Seems xgboost used quantile sketch, which is an approximated quantile algorithm anyways - the distributed version just added another "layer" to the approximate. I guess this may not be a big issue. We could just clarify that some difference for horizontal is expected due to feature quantile.

**Comment by trivialfis** [comment]
Will follow up tomorrow.

**Comment by trivialfis** [comment]
> If I understand correctly, xgboost used quantile sketch, which is an approximated quantile algorithm anyways - the distributed version just added another "layer" to the approximate. I guess this may not be a big issue. We could just clarify that some difference for horizontal is expected due to feature quantile.

100% correct. The diverge happens during the merge operation in the quantile sketch.

**Comment by ZiyueXu77** [comment]
> > If I understand correctly, xgboost used quantile sketch, which is an approximated quantile algorithm anyways - the distributed version just added another "layer" to the approximate. I guess this may not be a big issue. We could just clarify that some difference for horizontal is expected due to feature quantile.
> 
> 100% correct. The diverge happens during the merge operation in the quantile sketch.

Thanks! We can mark this issue solved :)
